src/svc_auth_gss.c

```diff
@@ -658,6 +658,7 @@ _svcauth_gss(struct svc_req *req, bool *no_dispatch)
 	if (rc != AUTH_OK) {
 		/* On success, the ref gets returned to the caller */
 		unref_svc_rpc_gss_data(gd);
+		req->rq_auth = NULL;
```

Comment:
Making req->rq_auth NULL unconditionally is incorrect. In case of gd->refcount ==1, then only mark it NULL as below / or similar to below logic:
if (rc != AUTH_OK) {
		/* On success, the ref gets returned to the caller */
		if(gd->refcnt ==1) req->rq_auth = NULL;
		unref_svc_rpc_gss_data(gd);
}

Reply:
I think @madhuthorat is right here.  We took a ref, and now we're dropping it, but we are still returning the ref'd item to the caller, who will try to drop the ref again.  We either need to null out rq_auth, or not drop the ref, and let the caller drop it.   I'm sort of leaning towards always returning with a ref, but I don't have a strong opinion.

@sachinpunadikar, can you explain why we should treat the ref == 1 case separately?

Reply:
@dang, My initial comment was in context of case for "RPCSEC_GSS_DATA". For this case we already increased refcnt for 2. In case of error condition, rc = !AUTH_OK, and as per new code rq_auth will be NULL. In such case we will be leaking rq_auth as it is with refcnt=1, and hence will not be released.

Reply:
I think in the case of RPCSEC_GSS_DATA, we should be okay.  In that case, we got it from the hash, and incremented the refcount by 1, and assigned it to this req.  In the case of an error, we should only dec it by one, not 2.  In fact, no mater what, we only incremented it by one in this function, so we should only decrement it by one, as well.
Am I missing something here?  Is there a case where we increment by 2?

Reply:
@dang and @madhuthorat, I rechecked the code. Your understanding is correct and the code changes are good to go.

Reply:
Thanks, Sachin.